ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/ExplicitDisclosureIT.scala

```diff
@@ -8,98 +8,218 @@ import com.daml.ledger.api.testtool.infrastructure.Allocation._
+      _ <- testContext.exerciseFetchDelegated(testContext.disclosedContract)
 
-      // Copy the actual Delegated contract (from the transaction stream of the owner) to a disclosed contract.
-      // Pretend we then send the disclosed contract to the delegate out of band.
-      disclosedContract = createEventToDisclosedContract(createDelegatedEvent)
+      // Create an extra disclosed contract
+      extraKey = ledger.nextKeyId()
+      _ <- ledger.create(owner, Delegated(owner, extraKey))
+      delegatedTxs <- ledger.flatTransactionsByTemplateId(Delegated.id, owner)
+      extraDelegatedEvent = createdEvents(delegatedTxs(1)).head
+      extraDisclosedContract = createEventToDisclosedContract(extraDelegatedEvent)
 
-      // Exercise a choice on the Delegation that fetches the Delegated contract
-      // Fails because the submitter doesn't see the contract being fetched
-      exerciseFetchRequest = fetchDelegatedRequest(delegationCid, delegatedCid, None)
-      exerciseFetchError <- ledger.submitAndWait(exerciseFetchRequest).failed
-      _ = assertGrpcError(
+      // Exercise the same choice, this time with superfluous disclosure
+      _ <- testContext.exerciseFetchDelegated(testContext.disclosedContract, extraDisclosedContract)
+    } yield {
+      assertGrpcError(
         exerciseFetchError,
         LedgerApiErrors.ConsistencyErrors.ContractNotFound,
         None,
         checkDefiniteAnswerMetadata = true,
       )
+    }
+  })
 
-      // Exercise the same choice, this time using correct explicit disclosure
-      exerciseFetchRequestWithDisclosure = fetchDelegatedRequest(
-        delegationCid,
-        delegatedCid,
-        Some(disclosedContract),
-      )
-      _ <- ledger.submitAndWait(exerciseFetchRequestWithDisclosure)
+  test(
+    "EDArchivedDisclosedContracts",
+    "The ledger rejects archived disclosed contracts",
+    allocate(Parties(2)),
+  )(implicit ec => { case Participants(Participant(ledger, owner, delegate)) =>
+    for {
+      testContext <- initializeTest(ledger, owner, delegate)
 
-      // Exercise the same choice, this time using bad explicit disclosure
-      malformedDisclosedContract = disclosedContract.copy(
-        arguments = Some(Delegated(owner, contractKey + "modified").arguments)
-      )
-      exerciseFetchRequestWithBadDisclosure = fetchDelegatedRequest(
-        delegationCid,
-        delegatedCid,
-        Some(malformedDisclosedContract),
+      // Archive the disclosed contract
+      _ <- ledger.exercise(owner, testContext.delegatedCid.exerciseArchive(_))
+
+      // Exercise the choice using the now inactive disclosed contract
+      exerciseError <- testContext.exerciseFetchDelegated(testContext.disclosedContract).failed
+    } yield {
+      // TODO DPP-1026: This should return ConsistencyErrors.ContractNotFound.
+      // Right now it fails to determine the LET because the algorithm doesn't exclude disclosed contracts from the LET lookup
```

Comment:
This is something that needs to be fixed in the PoC.

Reply:
Note that we shouldn’t just exclude them: We should use the let from the metadata for the LET check.